### PR TITLE
Add picocli in order to handle cli development

### PR DESCRIPTION
**Pull request by rmoreliovlabs**
## Description
Added picocli dependency plus an implementation wrapping one of our exisiting cli tools and making it compatible with old cli version.

## Motivation and Context
As of now, we use position based arguments in cli tools (eg. args[0], args[1] etc) and parse them “manually”, which doesn’t seem professional, is error prone and generally not flexible enough.

We could make use our existing configs based mechanism or some library (eg, [picocli](https://picocli.info/) - the one used in our benchmarking tool) to simplify arguments reading, parsing and documenting. With picocli we could even get --help option out of the box.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Requires Activation Code (Hard Fork)

* **Other information**:
PaowPeg: https://github.com/rsksmart/powpeg-node/pull/108

*fed:add-picocli*

### CODE REVIEW

File: rskj-core/src/main/java/co/rsk/cli/PicoCliToolRskContextAware.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+import java.util.concurrent.Callable;
+
+public abstract class PicoCliToolRskContextAware extends CliToolRskContextAware implements Callable<Integer> {
+    protected RskContext ctx;
+
+    @Override
+    protected void onExecute(@Nonnull String[] args, @Nonnull RskContext ctx) throws IOException {
+        this.ctx = ctx;
+
+        new CommandLine(this).setUnmatchedArgumentsAllowed(true).execute(args);
```
Comment (Vovchyk) [inline rskj-core/src/main/java/co/rsk/cli/PicoCliToolRskContextAware.java:34]: looks like the execution result of `CommandLine.execute()` is being ignored here

**Review by Vovchyk — CHANGES_REQUESTED** [review]

File: rskj-core/src/main/java/co/rsk/cli/tools/DbMigrate.java [hunk 2]
```diff
@@ -90,20 +94,19 @@ public static void main(String[] args) {
         create(MethodHandles.lookup().lookupClass()).execute(args);
     }
 
-    @SuppressWarnings("unused")
```
Comment (Vovchyk) [inline rskj-core/src/main/java/co/rsk/cli/tools/DbMigrate.java:93]: seems the ctor can be removed too

File: rskj-core/src/main/java/co/rsk/cli/CliToolRskContextAware.java [hunk 3]
```diff
@@ -74,6 +76,9 @@ public void execute(@Nonnull String[] args, @Nonnull Factory<RskContext> context
 
             printInfo("{} finished", cliToolName);
             stopper.stop(0);
+        }catch (PicocliBadResultException e) {
```
Comment (Vovchyk) [inline rskj-core/src/main/java/co/rsk/cli/CliToolRskContextAware.java:79]:
```suggestion
        } catch (PicocliBadResultException e) {
```

File: rskj-core/src/main/java/co/rsk/cli/PicoCliToolRskContextAware.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+import picocli.CommandLine;
+
+import javax.annotation.Nonnull;
+import java.io.IOException;
+import java.util.concurrent.Callable;
+
+public abstract class PicoCliToolRskContextAware extends CliToolRskContextAware implements Callable<Integer> {
+    protected RskContext ctx;
+
+    private static final Logger logger = LoggerFactory.getLogger(PicoCliToolRskContextAware.class);
```
Comment (Vovchyk) [inline rskj-core/src/main/java/co/rsk/cli/PicoCliToolRskContextAware.java:33]: seems the logger is not in use

File: rskj-core/src/main/java/co/rsk/cli/CliArgs.java [hunk 5]
```diff
@@ -100,7 +106,12 @@ public CliArgs<O, F> parse(String[] args) {
                             if (args[i].length() < 3) {
                                 throw new IllegalArgumentException("You must provide a flag name, e.g. --quiet");
                             }
-                            flags.add(getFlagByName(args[i].substring(2, args[i].length())));
+
+                            F f = getFlagByName(args[i].substring(2, args[i].length()));
```
Comment (Vovchyk) [inline rskj-core/src/main/java/co/rsk/cli/CliArgs.java:110]:
```suggestion
                            F flag = getFlagByName(args[i].substring(2, args[i].length()));
```

File: rskj-core/src/main/java/co/rsk/cli/CliArgs.java [hunk 6]
```diff
@@ -110,7 +121,13 @@ public CliArgs<O, F> parse(String[] args) {
                                         String.format("A value must be provided after the option -%s", args[i])
                                 );
                             }
-                            options.put(getOptionByName(args[i].substring(1, args[i].length())), args[i + 1]);
+
+                            O o = getOptionByName(args[i].substring(1, args[i].length()));
```
Comment (Vovchyk) [inline rskj-core/src/main/java/co/rsk/cli/CliArgs.java:125]:
```suggestion
                            O option = getOptionByName(args[i].substring(1, args[i].length()));
```

File: rskj-core/src/main/java/co/rsk/cli/CliArgs.java [hunk 7]
```diff
@@ -134,33 +151,42 @@ public CliArgs<O, F> parse(String[] args) {
         }
 
         private F getFlagByName(String flagName) {
-            return flags.stream()
+            F f = flags.stream()
```
Comment (Vovchyk) [inline rskj-core/src/main/java/co/rsk/cli/CliArgs.java:154]:
```suggestion
            F flag = flags.stream()
```

File: rskj-core/src/main/java/co/rsk/cli/CliArgs.java [hunk 8]
```diff
@@ -134,33 +151,42 @@ public CliArgs<O, F> parse(String[] args) {
...
+            if (f == null && !this.ignoreUnmatchedArgs) {
+                throw new NoSuchElementException(String.format("--%s is not a valid flag", flagName));
+            }
+
+            return f;
         }
 
         private O getOptionByName(String optionName) {
-            return options.stream()
+            O o = options.stream()
```
Comment (Vovchyk) [inline rskj-core/src/main/java/co/rsk/cli/CliArgs.java:167]:
```suggestion
            O option = options.stream()
```

**Review by Vovchyk — CHANGES_REQUESTED** [review]

File: rskj-core/src/main/java/co/rsk/cli/PicoCliToolRskContextAware.java (hunk 1 above)
Comment (rmoreliovlabs) [reply rskj-core/src/main/java/co/rsk/cli/PicoCliToolRskContextAware.java]: Solved

File: rskj-core/src/main/java/co/rsk/cli/CliArgs.java [hunk 9]
```diff
@@ -134,33 +151,42 @@ public CliArgs<O, F> parse(String[] args) {
...
          * Parses a string argument in the format e.g <i>database.dir=/home/rsk/core<i/> to a map in the following
          * structure:
          * <blockquote>
-         *     {
-         *         "database": {
-         *             "dir": "/home/rsk/core"
-         *         }
-         *     }
+         * {
+         * "database": {
```
[the remark this answers is not in the document]
Comment (rmoreliovlabs) [reply rskj-core/src/main/java/co/rsk/cli/CliArgs.java:184]: I didn't maybe intellij did when I wanted to format the code

File: rskj-core/src/main/java/co/rsk/cli/PicoCliToolRskContextAware.java (hunk 4 above)
Comment (rmoreliovlabs) [reply rskj-core/src/main/java/co/rsk/cli/PicoCliToolRskContextAware.java:33]: Done

File: rskj-core/src/main/java/co/rsk/cli/tools/DbMigrate.java (hunk 2 above)
Comment (rmoreliovlabs) [reply rskj-core/src/main/java/co/rsk/cli/tools/DbMigrate.java:93]: Done

File: rskj-core/src/main/java/co/rsk/cli/tools/DbMigrate.java [hunk 10]
```diff
@@ -45,10 +44,15 @@
...
  * - RocksDb (rocksdb as argument)
  */
-public class DbMigrate extends CliToolRskContextAware {
+@CommandLine.Command(name = "dbmigrate", mixinStandardHelpOptions = true, version = "dbmigrate 1.0",
+        description = "Migrates between different databases such as leveldb and rocksdb.")
+public class DbMigrate extends PicoCliToolRskContextAware {
     private static final Logger logger = LoggerFactory.getLogger(DbMigrate.class);
     private static final String NODE_ID_FILE = "nodeId.properties";
 
+    @CommandLine.Option(names = {"-tdb", "--targetdb"}, description = "The target db to migrate to. Example: leveldb, rocksdb ...")
```
Comment (Vovchyk) [inline rskj-core/src/main/java/co/rsk/cli/tools/DbMigrate.java:53]:
```suggestion
    @CommandLine.Option(names = {"-t", "--targetDb"}, description = "The target db to migrate to. Example: leveldb, rocksdb ...", required = true)
```

File: rskj-core/src/main/java/co/rsk/cli/tools/ExportBlocks.java [hunk 11]
```diff
@@ -39,26 +39,34 @@
  * - args[1] - to block number
  * - args[2] - file path
  */
-public class ExportBlocks extends CliToolRskContextAware {
+@CommandLine.Command(name = "exportblocks", mixinStandardHelpOptions = true, version = "exportblocks 1.0",
+        description = "The entry point for export blocks CLI tool")
```
Comment (Vovchyk) [inline rskj-core/src/main/java/co/rsk/cli/tools/ExportBlocks.java:43]:
```suggestion
        description = "Exports blocks from a specific block range to a file")
```

File: rskj-core/src/main/java/co/rsk/cli/tools/ExportBlocks.java [hunk 12]
```diff
@@ -39,26 +39,34 @@
...
+public class ExportBlocks extends PicoCliToolRskContextAware {
+    @CommandLine.Option(names = {"-fb", "--fromblock"}, description = "From block number", required = true)
```
Comment (Vovchyk) [inline rskj-core/src/main/java/co/rsk/cli/tools/ExportBlocks.java:45]:
```suggestion
    @CommandLine.Option(names = {"-fb", "--fromBlock"}, description = "From block number", required = true)
```

File: rskj-core/src/main/java/co/rsk/cli/tools/ExportBlocks.java [hunk 13]
```diff
@@ -39,26 +39,34 @@
...
+    private Long fromBlockNumber;
+
+    @CommandLine.Option(names = {"-tb", "--toblock"}, description = "To block number", required = true)
```
Comment (Vovchyk) [inline rskj-core/src/main/java/co/rsk/cli/tools/ExportBlocks.java:48]:
```suggestion
    @CommandLine.Option(names = {"-tb", "--toBlock"}, description = "To block number", required = true)
```

File: rskj-core/src/main/java/co/rsk/cli/tools/ExportBlocks.java [hunk 14]
```diff
@@ -39,26 +39,34 @@
...
+    private Long toBlockNumber;
+
+    @CommandLine.Option(names = {"-f", "--file"}, description = "File path", required = true)
```
Comment (Vovchyk) [inline rskj-core/src/main/java/co/rsk/cli/tools/ExportBlocks.java:51]:
```suggestion
    @CommandLine.Option(names = {"-f", "--file"}, description = "Path to a file to export blocks to", required = true)
```

File: rskj-core/src/main/java/co/rsk/cli/tools/ExportState.java [hunk 15]
```diff
@@ -41,26 +41,32 @@
  * - args[0] - block number
  * - args[1] - file path
  */
-public class ExportState extends CliToolRskContextAware {
+@CommandLine.Command(name = "exportstate", mixinStandardHelpOptions = true, version = "exportstate 1.0",
```
Comment (Vovchyk) [inline rskj-core/src/main/java/co/rsk/cli/tools/ExportState.java:44]:
```suggestion
@CommandLine.Command(name = "export-state", mixinStandardHelpOptions = true, version = "export-state 1.0",
```

File: rskj-core/src/main/java/co/rsk/cli/tools/ExportState.java [hunk 16]
```diff
@@ -41,26 +41,32 @@
...
+        description = "The entry point for export state CLI tool")
```
Comment (Vovchyk) [inline rskj-core/src/main/java/co/rsk/cli/tools/ExportState.java:45]: Exports state at specific block number to a file

File: rskj-core/src/main/java/co/rsk/cli/tools/ExportState.java [hunk 17]
```diff
@@ -41,26 +41,32 @@
...
+public class ExportState extends PicoCliToolRskContextAware {
+    @CommandLine.Option(names = {"-b", "--block"}, description = "Block number", required = true)
+    private Long blockNumber;
+
+    @CommandLine.Option(names = {"-f", "--file"}, description = "File path", required = true)
```
Comment (Vovchyk) [inline rskj-core/src/main/java/co/rsk/cli/tools/ExportState.java:50]:
```suggestion
    @CommandLine.Option(names = {"-f", "--file"}, description = "Path to a file to export state to", required = true)
```

File: rskj-core/src/main/java/co/rsk/cli/tools/ImportBlocks.java [hunk 18]
```diff
@@ -39,22 +38,27 @@
  * Required cli args:
  * - args[0] - file path
  */
-public class ImportBlocks extends CliToolRskContextAware {
+@CommandLine.Command(name = "importblocks", mixinStandardHelpOptions = true, version = "importblocks 1.0",
```
Comment (Vovchyk) [inline rskj-core/src/main/java/co/rsk/cli/tools/ImportBlocks.java:41]:
```suggestion
@CommandLine.Command(name = "import-blocks", mixinStandardHelpOptions = true, version = "import-blocks 1.0",
```

File: rskj-core/src/main/java/co/rsk/cli/tools/ImportBlocks.java [hunk 19]
```diff
@@ -39,22 +38,27 @@
...
+        description = "The entry point for import blocks CLI tool")
```
Comment (Vovchyk) [inline rskj-core/src/main/java/co/rsk/cli/tools/ImportBlocks.java:42]:
```suggestion
        description = "Imports blocks from a file")
```

File: rskj-core/src/main/java/co/rsk/cli/tools/ImportBlocks.java [hunk 20]
```diff
@@ -39,22 +38,27 @@
...
+public class ImportBlocks extends PicoCliToolRskContextAware {
+
+    @CommandLine.Option(names = {"-f", "--file"}, description = "File path", required = true)
```
Comment (Vovchyk) [inline rskj-core/src/main/java/co/rsk/cli/tools/ImportBlocks.java:45]:
```suggestion
    @CommandLine.Option(names = {"-f", "--file"}, description = "Path to a file to import blocks from", required = true)
```

File: rskj-core/src/main/java/co/rsk/cli/tools/ImportState.java [hunk 21]
```diff
@@ -39,26 +38,31 @@
  * Required cli args:
  * - args[0] - file path
  */
-public class ImportState extends CliToolRskContextAware {
+@CommandLine.Command(name = "importblocks", mixinStandardHelpOptions = true, version = "importblocks 1.0",
```
Comment (Vovchyk) [inline rskj-core/src/main/java/co/rsk/cli/tools/ImportState.java:41]:
```suggestion
@CommandLine.Command(name = "import-state", mixinStandardHelpOptions = true, version = "import-state 1.0",
```

File: rskj-core/src/main/java/co/rsk/cli/tools/ImportState.java [hunk 22]
```diff
@@ -39,26 +38,31 @@
...
+        description = "The entry point for import state CLI tool")
```
Comment (Vovchyk) [inline rskj-core/src/main/java/co/rsk/cli/tools/ImportState.java:42]:
```suggestion
        description = "Imports state from a file")
```

File: rskj-core/src/main/java/co/rsk/cli/tools/ImportState.java [hunk 23]
```diff
@@ -39,26 +38,31 @@
...
+public class ImportState extends PicoCliToolRskContextAware {
+    @CommandLine.Option(names = {"-f", "--file"}, description = "File path", required = true)
```
Comment (Vovchyk) [inline rskj-core/src/main/java/co/rsk/cli/tools/ImportState.java:44]:
```suggestion
    @CommandLine.Option(names = {"-f", "--file"}, description = "Path to a file to import state from", required = true)
```

File: rskj-core/src/main/java/co/rsk/cli/tools/IndexBlooms.java [hunk 24]
```diff
@@ -18,50 +18,79 @@
...
 import javax.annotation.Nonnull;
+import java.io.IOException;
+import java.util.concurrent.Callable;
 
 /**
  * The entry point for indexing block blooms
  * This is an experimental/unsupported tool
  */
-public class IndexBlooms {
+@CommandLine.Command(name = "indexblooms", mixinStandardHelpOptions = true, version = "indexblooms 1.0",
```
Comment (Vovchyk) [inline rskj-core/src/main/java/co/rsk/cli/tools/IndexBlooms.java:39]:
```suggestion
@CommandLine.Command(name = "index-blooms", mixinStandardHelpOptions = true, version = "index-blooms 1.0",
```

File: rskj-core/src/main/java/co/rsk/cli/tools/IndexBlooms.java [hunk 25]
```diff
@@ -18,50 +18,79 @@
...
+        description = "The entry point for indexing block blooms")
+public class IndexBlooms implements Callable<Integer> {
+
+    @CommandLine.Option(names = {"-fb", "--fromblock"}, description = "From block number", required = true)
```
Comment (Vovchyk) [inline rskj-core/src/main/java/co/rsk/cli/tools/IndexBlooms.java:43]:
```suggestion
    @CommandLine.Option(names = {"-fb", "--fromBlock"}, description = "From block number", required = true)
```

File: rskj-core/src/main/java/co/rsk/cli/tools/IndexBlooms.java [hunk 26]
```diff
@@ -18,50 +18,79 @@
...
+    private String fromBlockNumber;
+
+    @CommandLine.Option(names = {"-tb", "--toblock"}, description = "To block number", required = true)
```
Comment (Vovchyk) [inline rskj-core/src/main/java/co/rsk/cli/tools/IndexBlooms.java:46]:
```suggestion
    @CommandLine.Option(names = {"-tb", "--toBlock"}, description = "To block number", required = true)
```

File: rskj-core/src/main/java/co/rsk/cli/tools/RewindBlocks.java [hunk 27]
```diff
@@ -42,7 +43,11 @@
  * - "fmi" option can be used for finding minimum inconsistent block number and printing it to stdout. It'll print -1, if no such block is found;
  * - "rbc" option does two things: it looks for minimum inconsistent block and, if there's such, rewinds blocks from top one till the found one inclusively.
  */
-public class RewindBlocks extends CliToolRskContextAware {
+@CommandLine.Command(name = "rewindblocks", mixinStandardHelpOptions = true, version = "rewindblocks 1.0",
```
Comment (Vovchyk) [inline rskj-core/src/main/java/co/rsk/cli/tools/RewindBlocks.java:46]:
```suggestion
@CommandLine.Command(name = "rewind-blocks", mixinStandardHelpOptions = true, version = "rewind-blocks 1.0",
```

File: rskj-core/src/main/java/co/rsk/cli/tools/StartBootstrap.java [hunk 28]
```diff
@@ -39,19 +43,38 @@
  *
  * Note: this is an experimental tool
  */
-public class StartBootstrap {
+@CommandLine.Command(name = "startbootstrap", mixinStandardHelpOptions = true, version = "startbootstrap 1.0",
```
Comment (Vovchyk) [inline rskj-core/src/main/java/co/rsk/cli/tools/StartBootstrap.java:46]:
```suggestion
@CommandLine.Command(name = "start-bootstrap", mixinStandardHelpOptions = true, version = "start-bootstrap 1.0",
```

File: rskj-core/src/main/java/co/rsk/cli/tools/StartBootstrap.java [hunk 29]
```diff
@@ -39,19 +43,38 @@
...
+        description = "The bootstrap node starts one service which only participates in the peer discovery protocol")
```
Comment (Vovchyk) [inline rskj-core/src/main/java/co/rsk/cli/tools/StartBootstrap.java:47]:
```suggestion
        description = "Starts a bootstrap node with one service, which only participates in the peer discovery protocol")
```

**Review by Vovchyk — CHANGES_REQUESTED** [review]

**Review by Vovchyk — CHANGES_REQUESTED** [review]
this may require a counterpart PR for `powpeg`

**Comment by rmoreliovlabs** [comment]
pipeline:run

**Comment by rmoreliovlabs** [comment]
pipeline:run

**Comment by Vovchyk** [comment]
pipeline:run

**Comment by rmoreliovlabs** [comment]
pipeline:run

**Comment by rmoreliovlabs** [comment]
pipeline:run

**Comment by Vovchyk** [comment]
pipeline:run

**Comment by rmoreliovlabs** [comment]
pipeline:run

**Comment by rmoreliovlabs** [comment]
pipelin:run

**Comment by rmoreliovlabs** [comment]
pipeline:run

**Comment by Vovchyk** [comment]
pipeline:run

**Comment by rmoreliovlabs** [comment]
pipeline:run

**Comment by Vovchyk** [comment]
pipeline:run

**Comment by Vovchyk** [comment]
pipeline:run

**Comment by Vovchyk** [comment]
pipeline:run

**Comment by rmoreliovlabs** [comment]
pipeline:run

**Review by Vovchyk — DISMISSED** [review]

**Comment by Vovchyk** [comment]
pipeline:run

**Comment by Vovchyk** [comment]
pipeline:run

**Comment by Vovchyk** [comment]
pipeline:run

**Comment by Vovchyk** [comment]
pipeline:run

**Comment by rmoreliovlabs** [comment]
pipeline:run

**Review by Vovchyk — APPROVED** [review]

**Pull request merged**